File: packages/backend/app/routes/groups.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from ..database import get_connection
from ..auth import require_auth

router = APIRouter(prefix="/groups", tags=["Groups"])

@router.get("/filters")
async def get_filters():
    pool = await get_connection()
    
    teachers_rows = await pool.fetch(
        """
        SELECT DISTINCT u.id, u.name
        FROM users u
        INNER JOIN teachers t ON t.user_id = u.id
        WHERE EXISTS (
            SELECT 1 FROM groups g WHERE g.main_teacher_id = t.id AND g.is_closed = FALSE
        ) OR EXISTS (
            SELECT 1 FROM group_teachers gt
            INNER JOIN groups g ON g.id = gt.group_id
            WHERE gt.teacher_id = t.id AND g.is_closed = FALSE
        )
        ORDER BY u.name
        """
    )
    
    halls_rows = await pool.fetch(
        """
        SELECT DISTINCT h.id, h.name
        FROM halls h
        WHERE EXISTS (
            SELECT 1 FROM groups g WHERE g.hall_id = h.id AND g.is_closed = FALSE
        )
        ORDER BY h.name
        """
    )
    
    return {
        "teachers": [{"id": r["id"], "name": r["name"]} for r in teachers_rows],
        "halls": [{"id": r["id"], "name": r["name"]} for r in halls_rows]
    }

@router.get("/available")
async def get_available_groups():
    pool = await get_connection()
    
    rows = await pool.fetch(
        """
        SELECT
            g.id,
            g.name,
            g.capacity,
            g.start_time,
            g.duration_minutes,
            h.id AS hall_id,
            h.name AS hall_name,
            h.capacity AS hall_capacity,
            (SELECT COUNT(*) FROM group_students gs WHERE gs.group_id = g.id AND gs.is_trial = FALSE) AS enrolled,
            array_remove(array_agg(DISTINCT COALESCE(gt.teacher_id, g.main_teacher_id)), NULL) AS teacher_ids,
            array_remove(array_agg(DISTINCT u.name), NULL) AS teacher_names
        FROM groups g
        LEFT JOIN halls h ON h.id = g.hall_id
        LEFT JOIN group_teachers gt ON gt.group_id = g.id
        LEFT JOIN teachers t ON t.id = COALESCE(gt.teacher_id, g.main_teacher_id)
        LEFT JOIN users u ON u.id = t.user_id
        WHERE g.is_closed = FALSE AND g.is_additional = FALSE
        GROUP BY g.id, g.name, g.capacity, g.start_time, g.duration_minutes, h.id, h.name, h.capacity
        ORDER BY g.start_time NULLS LAST, g.id
        """
    )
    
    response = []
    for row in rows:
        capacity = row["capacity"]
        enrolled = int(row["enrolled"]) if row["enrolled"] else 0
        free_slots = capacity - enrolled if capacity else None
        
        teacher_names = row["teacher_names"] or []
        teacher_ids = row["teacher_ids"] or []
        
        response.append({
            "id": row["id"],
            "name": row["name"],
            "capacity": capacity,
            "start_time": str(row["start_time"]) if row["start_time"] else None,
            "duration_minutes": row["duration_minutes"],
            "hall": {
                "id": row["hall_id"],
                "name": row["hall_name"],
                "capacity": row["hall_capacity"]
            } if row["hall_id"] else None,
            "enrolled": enrolled,
            "free_slots": free_slots,
            "teacher_ids": [int(t) for t in teacher_ids if t],
            "teacher_names": [str(t) for t in teacher_names if t]
        })
    
    return response

@router.get("/schedule")
async def get_schedule(hallId: Optional[int] = None):
    pool = await get_connection()
    
    if hallId:
        base_schedule = await pool.fetch(
            """
            SELECT g.id, g.name, g.start_time, g.duration_minutes, g.is_additional,
                   h.id AS hall_id, h.name AS hall_name
            FROM groups g
            LEFT JOIN halls h ON h.id = g.hall_id
            WHERE g.hall_id = $1
            ORDER BY g.start_time NULLS LAST, g.id
            """,
            hallId
        )
        
        additions = await pool.fetch(
            """
            SELECT se.id, se.group_id, se.start_time, se.duration_minutes, se.additional,
                   se.approved, se.requested_by_student, se.reason, h.id, h.name
            FROM schedule_exceptions se
            LEFT JOIN halls h ON h.id = se.hall_id
            WHERE se.hall_id = $1
            ORDER BY se.start_time NULLS LAST, se.id
            """,
            hallId
        )
    else:
        base_schedule = await pool.fetch(
            """
            SELECT g.id, g.name, g.start_time, g.duration_minutes, g.is_additional,
                   h.id AS hall_id, h.name AS hall_name
            FROM groups g
            LEFT JOIN halls h ON h.id = g.hall_id
            ORDER BY g.start_time NULLS LAST, g.id
            """
        )
        
        additions = await pool.fetch(
            """
            SELECT se.id, se.group_id, se.start_time, se.duration_minutes, se.additional,
                   se.approved, se.requested_by_student, se.reason, h.id, h.name
            FROM schedule_exceptions se
            LEFT JOIN halls h ON h.id = se.hall_id
            ORDER BY se.start_time NULLS LAST, se.id
            """
        )
    
    additions_by_group = {}
    for row in additions:
        group_id = row["group_id"]
        if group_id not in additions_by_group:
            additions_by_group[group_id] = []
        additions_by_group[group_id].append({
            "id": row["id"],
            "start_time": str(row["start_time"]) if row["start_time"] else None,
            "duration_minutes": row["duration_minutes"],
            "is_additional": row["additional"],
            "approved": row["approved"],
            "requested_by_student": row["requested_by_student"],
            "reason": row["reason"],
            "hall": {"id": row[8], "name": row[9]} if row[8] else None
        })
    
    result = []
    for row in base_schedule:
        group_id = row["id"]
        result.append({
            "group_id": group_id,
            "group_name": row["name"],
            "start_time": str(row["start_time"]) if row["start_time"] else None,
            "duration_minutes": row["duration_minutes"],
            "is_additional": row["is_additional"],
            "hall": {"id": row["hall_id"], "name": row["hall_name"]} if row["hall_id"] else None,
            "exceptions": additions_by_group.get(group_id, [])
        })
    
    return result

async def resolve_student_id(pool, user_id: int) -> Optional[int]:
    row = await pool.fetchrow(
        "SELECT id FROM students WHERE user_id = $1",
        user_id
    )
    return row["id"] if row else None

@router.post("/{group_id}/join")
async def join_group(group_id: int, user: dict = Depends(require_auth)):
    if user.get("role") != "student":
        raise HTTPException(status_code=403, detail="Only students may join groups")
    
    pool = await get_connection()
    student_id = await resolve_student_id(pool, user["id"])
    
    if not student_id:
        raise HTTPException(status_code=404, detail="Student profile not found")
    
    info = await pool.fetchrow(
        """
        SELECT
            capacity,
            is_closed,
            (SELECT COUNT(*) FROM group_students gs WHERE gs.group_id = $1 AND gs.is_trial = FALSE) AS enrolled,
            EXISTS(
                SELECT 1 FROM group_students gs WHERE gs.group_id = $1 AND gs.student_id = $2
            ) AS already_joined
        FROM groups
        WHERE id = $1
        """,
        group_id, student_id
    )
    
    if not info:
        raise HTTPException(status_code=404, detail="Group not found")
    
    if info["is_closed"]:
        raise HTTPException(status_code=409, detail="Group is closed for enrollment")
    
    if info["already_joined"]:
        raise HTTPException(status_code=409, detail="Already joined this group")
    
    capacity = info["capacity"]
    enrolled = int(info["enrolled"]) if info["enrolled"] else 0
    
    if capacity and enrolled >= capacity:
        raise HTTPException(status_code=409, detail="Group is full")
    
    await pool.execute(
        """
        INSERT INTO group_students (group_id, student_id, is_trial)
        VALUES ($1, $2, FALSE)
        ON CONFLICT DO NOTHING
        """,
        group_id, student_id
    )
    
    return {"message": "Successfully joined the group"}

@router.post("/{group_id}/trial")
async def trial_lesson(group_id: int, user: dict = Depends(require_auth)):
    if user.get("role") != "student":
        raise HTTPException(status_code=403, detail="Only students may request trial lessons")
    
    pool = await get_connection()
    student_id = await resolve_student_id(pool, user["id"])
    
    if not student_id:
        raise HTTPException(status_code=404, detail="Student profile not found")
    
    student = await pool.fetchrow(
        "SELECT trial_used FROM students WHERE id = $1",
        student_id
    )
    
    if student and student["trial_used"]:
        raise HTTPException(status_code=409, detail="Trial lesson already used")
    
    await pool.execute(
        """
        INSERT INTO group_students (group_id, student_id, is_trial)
        VALUES ($1, $2, TRUE)
        ON CONFLICT DO NOTHING
        """,
        group_id, student_id
    )
    
    await pool.execute(
        "UPDATE students SET trial_used = TRUE WHERE id = $1",
        student_id
    )
    
    return {"message": "Trial lesson registered"}

class AdditionalLessonRequest(BaseModel):
    reason: Optional[str] = None

@router.post("/{group_id}/additional-request")
async def request_additional_lesson(group_id: int, data: AdditionalLessonRequest, user: dict = Depends(require_auth)):
    if user.get("role") != "student":
        raise HTTPException(status_code=403, detail="Only students may request additional lessons")
    
    pool = await get_connection()
    student_id = await resolve_student_id(pool, user["id"])
    
    if not student_id:
        raise HTTPException(status_code=404, detail="Student profile not found")
    
    is_member = await pool.fetchrow(
        "SELECT 1 FROM group_students WHERE group_id = $1 AND student_id = $2",
        group_id, student_id
    )
    
    if not is_member:
        raise HTTPException(status_code=403, detail="Must be a member of the group")
    
    group_info = await pool.fetchrow(
        "SELECT hall_id, main_teacher_id FROM groups WHERE id = $1",
        group_id
    )
    
    if not group_info:
        raise HTTPException(status_code=404, detail="Group not found")
    
    result = await pool.fetchrow(
        """
        INSERT INTO schedule_exceptions (
            group_id, teacher_id, hall_id, start_time, duration_minutes,
            reason, additional, approved, requested_by_student
        ) VALUES ($1, $2, $3, NOW(), 90, $4, TRUE, FALSE, TRUE)
        RETURNING id
        """,
        group_id, group_info["main_teacher_id"], group_info["hall_id"], data.reason
    )
    
    return {"request_id": result["id"], "message": "Additional lesson request submitted"}

File: packages/frontend/nextjs/lib/routes/groups.dart
import 'dart:convert';

import 'package:postgres/postgres.dart' show Sql;
import 'package:shelf/shelf.dart';
import 'package:shelf_router/shelf_router.dart';

import '../db/connection.dart';

class GroupsRoute {
  Router get router {
    final router = Router();

    router.get('/groups/available', _availableGroups);
    router.get('/groups/schedule', _groupSchedule);
    router.get('/groups/filters', _getFilters);
    router.post('/groups/<id>/join', _joinGroup);
    router.post('/groups/<id>/trial', _trialLesson);
    router.post('/groups/<id>/additional-request', _requestAdditionalLesson);

    return router;
  }

  Future<Response> _getFilters(Request req) async {
    try {
      
      final teachersRows = await connection.execute('''
        SELECT DISTINCT u.id, u.name
        FROM users u
        INNER JOIN teachers t ON t.user_id = u.id
        WHERE EXISTS (
          SELECT 1 FROM groups g WHERE g.main_teacher_id = t.id AND g.is_closed = FALSE
        ) OR EXISTS (
          SELECT 1 FROM group_teachers gt
          INNER JOIN groups g ON g.id = gt.group_id
          WHERE gt.teacher_id = t.id AND g.is_closed = FALSE
        )
        ORDER BY u.name;
      ''');

      
      final hallsRows = await connection.execute('''
        SELECT DISTINCT h.id, h.name
        FROM halls h
        WHERE EXISTS (
          SELECT 1 FROM groups g WHERE g.hall_id = h.id AND g.is_closed = FALSE
        )
        ORDER BY h.name;
      ''');

      final teachers = teachersRows.map((row) => {
        'id': row[0],
        'name': row[1],
      }).toList();

      final halls = hallsRows.map((row) => {
        'id': row[0],
        'name': row[1],
      }).toList();

      return Response.ok(
        jsonEncode({
          'teachers': teachers,
          'halls': halls,
        }),
        headers: {'content-type': 'application/json'},
      );
    } catch (e) {
      print('❌ Failed to load filters: $e');
      return Response.internalServerError(
        body: jsonEncode({'error': 'Failed to load filters'}),
        headers: {'content-type': 'application/json'},
      );
    }
  }

  Future<Response> _availableGroups(Request req) async {
    try {
      final rows = await connection.execute('''
        SELECT
          g.id,
          g.name,
          g.capacity,
          g.start_time,
          g.duration_minutes,
          h.id AS hall_id,
          h.name AS hall_name,
          h.capacity AS hall_capacity,
          (SELECT COUNT(*) FROM group_students gs WHERE gs.group_id = g.id AND gs.is_trial = FALSE) AS enrolled,
          array_remove(array_agg(DISTINCT COALESCE(gt.teacher_id, g.main_teacher_id)), NULL) AS teacher_ids,
          array_remove(array_agg(DISTINCT u.name), NULL) AS teacher_names
        FROM groups g
        LEFT JOIN halls h ON h.id = g.hall_id
        LEFT JOIN group_teachers gt ON gt.group_id = g.id
        LEFT JOIN teachers t ON t.id = COALESCE(gt.teacher_id, g.main_teacher_id)
        LEFT JOIN users u ON u.id = t.user_id
        WHERE g.is_closed = FALSE AND g.is_additional = FALSE
        GROUP BY g.id, g.name, g.capacity, g.start_time, g.duration_minutes, h.id, h.name, h.capacity
        ORDER BY g.start_time NULLS LAST, g.id;
      ''');

      final response = rows
          .map((row) {
            final capacity = row[2] as int?;
            final enrolled = row[8] as int;
            final freeSlots = capacity == null ? null : (capacity - enrolled);
            final teacherNames = (row[10] as List?)?.whereType<String>().toList() ?? const <String>[];
            return {
              'id': row[0],
              'name': row[1],
              'capacity': capacity,
              'start_time': row[3]?.toString(),
              'duration_minutes': row[4],
              'hall': row[5] == null
                  ? null
                  : {
                      'id': row[5],
                      'name': row[6],
                      'capacity': row[7],
                    },
              'enrolled': enrolled,
              
              'free_slots': freeSlots,
              'teacher_ids': (row[9] as List?)?.whereType<int>().toList() ?? const <int>[],
              'teacher_names': teacherNames,
            };
          }).toList();

      return Response.ok(
        jsonEncode(response),
        headers: {'content-type': 'application/json'},
      );
    } catch (e) {
      print('❌ Failed to load available groups: $e');
      return Response.internalServerError(
        body: jsonEncode({'error': 'Failed to load groups'}),
        headers: {'content-type': 'application/json'},
      );
    }
  }

  Future<Response> _groupSchedule(Request req) async {
    final hallIdParam = req.url.queryParameters['hallId'];
    int? hallId;
    if (hallIdParam != null) {
      hallId = int.tryParse(hallIdParam);
      if (hallId == null) {
        return Response(
          400,
          body: jsonEncode({'error': 'hallId must be an integer'}),
          headers: {'content-type': 'application/json'},
        );
      }
    }

    try {
      final baseSchedule = await connection.execute(
        Sql.named('''
          SELECT g.id, g.name, g.start_time, g.duration_minutes, g.is_additional,
                 h.id AS hall_id, h.name AS hall_name
          FROM groups g
          LEFT JOIN halls h ON h.id = g.hall_id
          WHERE (@hallId::INT IS NULL OR g.hall_id = @hallId)
          ORDER BY g.start_time NULLS LAST, g.id;
        '''),
        parameters: {'hallId': hallId},
      );

      final additions = await connection.execute(
        Sql.named('''
          SELECT se.id, se.group_id, se.start_time, se.duration_minutes, se.additional,
                 se.approved, se.requested_by_student, se.reason, h.id, h.name
          FROM schedule_exceptions se
          LEFT JOIN halls h ON h.id = se.hall_id
          WHERE (@hallId::INT IS NULL OR se.hall_id = @hallId)
          ORDER BY se.start_time NULLS LAST, se.id;
        '''),
        parameters: {'hallId': hallId},
      );

      final additionsByGroup = <int, List<Map<String, dynamic>>>{};
      for (final row in additions) {
        final groupId = row[1] as int;
        additionsByGroup.putIfAbsent(groupId, () => []).add({
              'id': row[0],
              'start_time': row[2]?.toString(),
              'duration_minutes': row[3],
              'is_additional': row[4],
              'approved': row[5],
              'requested_by_student': row[6],
              'reason': row[7],
              'hall': row[8] == null
                  ? null
                  : {
                      'id': row[8],
                      'name': row[9],
                    },
            });
      }

      final result = baseSchedule
          .map((row) => {
                'group_id': row[0],
                'group_name': row[1],
                'start_time': row[2]?.toString(),
                'duration_minutes': row[3],
                'is_additional': row[4],
                'hall': row[5] == null
                    ? null
                    : {
                        'id': row[5],
                        'name': row[6],
                      },
                'exceptions': additionsByGroup[row[0]] ?? const [],
              })
          .toList();

      return Response.ok(
        jsonEncode(result),
        headers: {'content-type': 'application/json'},
      );
    } catch (e) {
      print('❌ Failed to load schedule: $e');
      return Response.internalServerError(
        body: jsonEncode({'error': 'Failed to load schedule'}),
        headers: {'content-type': 'application/json'},
      );
    }
  }

  Future<Response> _joinGroup(Request req, String id) async {
    final user = req.context['user'] as Map<String, dynamic>?;
    if (user == null || user['role'] != 'student') {
      return Response.forbidden(
        jsonEncode({'error': 'Only students may join groups'}),
        headers: {'content-type': 'application/json'},
      );
    }

    final groupId = int.tryParse(id);
    if (groupId == null) {
      return Response(
        400,
        body: jsonEncode({'error': 'Invalid group id'}),
        headers: {'content-type': 'application/json'},
      );
    }

    try {
      final studentId = await _resolveStudentId(user['id'] as int);
      if (studentId == null) {
        return Response(
          404,
          body: jsonEncode({'error': 'Student profile not found'}),
          headers: {'content-type': 'application/json'},
        );
      }

      final info = await connection.execute(
        Sql.named('''
          SELECT
            capacity,
            is_closed,
            (SELECT COUNT(*) FROM group_students gs WHERE gs.group_id = @groupId AND gs.is_trial = FALSE) AS enrolled,
            EXISTS(
              SELECT 1 FROM group_students gs WHERE gs.group_id = @groupId AND gs.student_id = @studentId
            ) AS already_joined
          FROM groups
          WHERE id = @groupId;
        '''),
        parameters: {'groupId': groupId, 'studentId': studentId},
      );

      if (info.isEmpty) {
        return Response(
          404,
          body: jsonEncode({'error': 'Group not found'}),
          headers: {'content-type': 'application/json'},
        );
      }

      final row = info.first;
      final capacity = row[0] as int;
      final isClosed = row[1] as bool;
      final enrolled = row[2] as int;
      final alreadyJoined = row[3] as bool;

      if (isClosed) {
        return Response(
          409,
          body: jsonEncode({'error': 'Group is closed for enrollment'}),
          headers: {'content-type': 'application/json'},
        );
      }
      
      final groupTime = await connection.execute(
        Sql.named('SELECT start_time, duration_minutes FROM groups WHERE id = @groupId'),
        parameters: {'groupId': groupId},
      );
      if (groupTime.isNotEmpty && groupTime.first[0] != null) {
        final gStart = groupTime.first[0] as DateTime;
        final gDuration = groupTime.first[1] as int? ?? 90;
        final gEnd = gStart.add(Duration(minutes: gDuration));

        final overlap = await connection.execute(
          Sql.named('''
            SELECT gs.group_id FROM group_students gs
            JOIN groups g ON g.id = gs.group_id
            WHERE gs.student_id = @studentId
              AND gs.group_id != @groupId
              AND NOT (
                (g.start_time + (g.duration_minutes * INTERVAL '1 minute')) <= @gStart
                OR g.start_time >= @gEnd
              )
            LIMIT 1;
          '''),
          parameters: {'studentId': studentId, 'groupId': groupId, 'gStart': gStart, 'gEnd': gEnd},
        );

        if (overlap.isNotEmpty) {
          return Response(
            409,
            body: jsonEncode({'error': 'Student has another enrolled group that overlaps in time'}),
            headers: {'content-type': 'application/json'},
          );
        }
      }
      if (alreadyJoined) {
        return Response(
          409,
          body: jsonEncode({'error': 'Student already joined this group'}),
          headers: {'content-type': 'application/json'},
        );
      }
      if (enrolled >= capacity) {
        return Response(
          409,
          body: jsonEncode({'error': 'Group capacity reached'}),
          headers: {'content-type': 'application/json'},
        );
      }

      await connection.execute(
        Sql.named('''
          INSERT INTO group_students (group_id, student_id, is_trial)
          VALUES (@groupId, @studentId, FALSE)
          ON CONFLICT (group_id, student_id) DO UPDATE SET is_trial = FALSE;
        '''),
        parameters: {'groupId': groupId, 'studentId': studentId},
      );

      print('👤 Student $studentId joined group $groupId');
      return Response.ok(
        jsonEncode({'message': 'Student enrolled to group'}),
        headers: {'content-type': 'application/json'},
      );
    } catch (e) {
      print('❌ Failed to join group: $e');
      return Response.internalServerError(
        body: jsonEncode({'error': 'Unable to join group'}),
        headers: {'content-type': 'application/json'},
      );
    }
  }

  Future<Response> _trialLesson(Request req, String id) async {
    final user = req.context['user'] as Map<String, dynamic>?;
    if (user == null || user['role'] != 'student') {
      return Response.forbidden(
        jsonEncode({'error': 'Only students may request a trial lesson'}),
        headers: {'content-type': 'application/json'},
      );
    }

    final groupId = int.tryParse(id);
    if (groupId == null) {
      return Response(
        400,
        body: jsonEncode({'error': 'Invalid group id'}),
        headers: {'content-type': 'application/json'},
      );
    }

    try {
      final studentId = await _resolveStudentId(user['id'] as int);
      if (studentId == null) {
        return Response(
          404,
          body: jsonEncode({'error': 'Student profile not found'}),
          headers: {'content-type': 'application/json'},
        );
      }

      final eligibility = await connection.execute(
        Sql.named('''
          SELECT s.trial_used,
                 EXISTS(
                   SELECT 1 FROM group_students gs WHERE gs.group_id = @groupId AND gs.student_id = s.id
                 ) AS already_joined,
                 g.capacity,
                 g.is_closed,
                 (SELECT COUNT(*) FROM group_students gs WHERE gs.group_id = @groupId AND gs.is_trial = FALSE) AS enrolled
          FROM students s
          JOIN groups g ON g.id = @groupId
          WHERE s.id = @studentId;
        '''),
        parameters: {'groupId': groupId, 'studentId': studentId},
      );

      if (eligibility.isEmpty) {
        return Response(
          404,
          body: jsonEncode({'error': 'Student or group not found'}),
          headers: {'content-type': 'application/json'},
        );
      }

      final row = eligibility.first;
      final trialUsed = row[0] as bool;
      final alreadyJoined = row[1] as bool;
      final capacity = row[2] as int;
      final isClosed = row[3] as bool;
      final enrolled = row[4] as int;

      if (trialUsed) {
        return Response(
          409,
          body: jsonEncode({'error': 'Trial lesson already used'}),
          headers: {'content-type': 'application/json'},
        );
      }
      if (alreadyJoined) {
        return Response(
          409,
          body: jsonEncode({'error': 'Student already assigned to this group'}),
          headers: {'content-type': 'application/json'},
        );
      }
      if (isClosed || enrolled >= capacity) {
        return Response(
          409,
          body: jsonEncode({'error': 'Group unavailable for trials'}),
          headers: {'content-type': 'application/json'},
        );
      }

      await connection.execute(
        Sql.named('''
          INSERT INTO group_students (group_id, student_id, is_trial)
          VALUES (@groupId, @studentId, TRUE)
          ON CONFLICT (group_id, student_id) DO UPDATE SET is_trial = TRUE;
        '''),
        parameters: {'groupId': groupId, 'studentId': studentId},
      );
      await connection.execute(
        Sql.named('UPDATE students SET trial_used = TRUE WHERE id = @studentId'),
        parameters: {'studentId': studentId},
      );

      print('🎟️ Student $studentId booked trial for group $groupId');
      return Response.ok(
        jsonEncode({'message': 'Trial lesson booked'}),
        headers: {'content-type': 'application/json'},
      );
    } catch (e) {
      print('❌ Failed to book trial: $e');
      return Response.internalServerError(
        body: jsonEncode({'error': 'Unable to book trial'}),
        headers: {'content-type': 'application/json'},
      );
    }
  }

  Future<Response> _requestAdditionalLesson(Request req, String id) async {
    final user = req.context['user'] as Map<String, dynamic>?;
    if (user == null || user['role'] != 'student') {
      return Response.forbidden(
        jsonEncode({'error': 'Only students may request additional lessons'}),
        headers: {'content-type': 'application/json'},
      );
    }

    final groupId = int.tryParse(id);
    if (groupId == null) {
      return Response(
        400,
        body: jsonEncode({'error': 'Invalid group id'}),
        headers: {'content-type': 'application/json'},
      );
    }

    try {
      final body = await req.readAsString();
      final data = body.isEmpty ? <String, dynamic>{} : jsonDecode(body) as Map<String, dynamic>;
      final requestedStart = DateTime.tryParse(data['start_time']?.toString() ?? '');
      final reason = data['reason']?.toString();

      if (requestedStart == null) {
        return Response(
          400,
          body: jsonEncode({'error': 'start_time is required in ISO8601 format'}),
          headers: {'content-type': 'application/json'},
        );
      }

      final studentId = await _resolveStudentId(user['id'] as int);
      if (studentId == null) {
        return Response(
          404,
          body: jsonEncode({'error': 'Student profile not found'}),
          headers: {'content-type': 'application/json'},
        );
      }

      final groupInfo = await connection.execute(
        Sql.named('SELECT hall_id, duration_minutes FROM groups WHERE id = @groupId'),
        parameters: {'groupId': groupId},
      );

      if (groupInfo.isEmpty) {
        return Response(
          404,
          body: jsonEncode({'error': 'Group not found'}),
          headers: {'content-type': 'application/json'},
        );
      }

      await connection.execute(
        Sql.named('''
          INSERT INTO schedule_exceptions (
            group_id,
            teacher_id,
            hall_id,
            start_time,
            duration_minutes,
            reason,
            additional,
            approved,
            requested_by_student
          )
          VALUES (
            @groupId,
            NULL,
            @hallId,
            @startTime,
            @duration,
            @reason,
            TRUE,
            FALSE,
            TRUE
          );
        '''),
        parameters: {
          'groupId': groupId,
          'hallId': groupInfo.first[0],
          'startTime': requestedStart,
          'duration': groupInfo.first[1],
          'reason': reason,
        },
      );

      print('📝 Student $studentId requested additional lesson for group $groupId');
      return Response.ok(
        jsonEncode({'message': 'Additional lesson request submitted for review'}),
        headers: {'content-type': 'application/json'},
      );
    } catch (e) {
      print('❌ Failed to request additional lesson: $e');
      return Response.internalServerError(
        body: jsonEncode({'error': 'Unable to submit additional lesson request'}),
        headers: {'content-type': 'application/json'},
      );
    }
  }

  Future<int?> _resolveStudentId(int userId) async {
    final rows = await connection.execute(
      Sql.named('SELECT id FROM students WHERE user_id = @userId'),
      parameters: {'userId': userId},
    );
    if (rows.isEmpty) {
      return null;
    }
    return rows.first[0] as int;
  }
}
